### ISSUE TITLE
Unable to build with cmake FetchContent

### ISSUE THREAD

**Issue by kagurazaka-ayano** [AnswerDotAI/gpu.cpp#42]
Expected: Use FetchContent_Delcare to declare the dependency and then use FetchContent_MakeAvailable and the built library and headers will be in the _deps/gpucpp-{build/subbuild/src}

Actual: The generating process ends up with an error. Error message and minimal reproducible CMakeLists.txt is inside the [gist](https://gist.github.com/kagurazaka-ayano/64a806bff6de0c6a7ba7a00d12eff3dc)

The dawn library file doesn't appear in the third_party/lib directory. After I built [this](https://github.com/jspanchu/webgpu-dawn-binaries) and put the libdawn.dylib into the third_party/lib directory the cmake command works.

**Comment by austinvhuang** [comment]
tagging @MichealReed as well - it's been a minute since i've played with cmake builds.  From a quick look don't we need to include dawn either/both transitive fetchcontent dependence or make use of the prebuilt binaries:

- incorporate functionality to download the prebuilt dylib similarly to the makefile + setup.py script
- transitive via fetchcontent like https://github.com/AnswerDotAI/gpu.cpp/blob/main/examples/webgpu_from_scratch/CMakeLists.txt 
- transitive shared library with fetchcontent. + jspanchu repo

can take a closer look over the weekend. would probably good if we have a separate template starter repo in the future if anyone's interesting in taking a stab at that.

**Comment by kagurazaka-ayano** [comment]
> tagging @MichealReed as well - it's been a minute since i've played with cmake builds. From a quick look I think we need to include dawn either/both transitive fetchcontent dependence or make use of the prebuilt binaries:
> 
> * incorporate functionality to download the prebuilt dylib similarly to the makefile + setup.py script
> * transitive via fetchcontent like https://github.com/AnswerDotAI/gpu.cpp/blob/main/examples/webgpu_from_scratch/CMakeLists.txt
> * transitive shared library with fetchcontent. + jspanchu repo
> 
> can take a closer look over the weekend. would probably good if we have a separate template starter repo in the future if anyone's interesting in taking a stab at that.

Downloading prebuilt library can be done in FetchContent since it accepts url.

**Comment by MichealReed** [comment]
> tagging @MichealReed as well - it's been a minute since i've played with cmake builds. From a quick look I think we need to include dawn either/both transitive fetchcontent dependence or make use of the prebuilt binaries:
> 
>     * incorporate functionality to download the prebuilt dylib similarly to the makefile + setup.py script
> 
>     * transitive via fetchcontent like https://github.com/AnswerDotAI/gpu.cpp/blob/main/examples/webgpu_from_scratch/CMakeLists.txt
> 
>     * transitive shared library with fetchcontent. + jspanchu repo
> 
> 
> can take a closer look over the weekend. would probably good if we have a separate template starter repo in the future if anyone's interesting in taking a stab at that.

Yes this makes sense, our makefile was the component that downloaded the libs before calling cmake. Nothing is in CMake currently to download libs and the python script must be ran first. Adding it via fetch seems like a good addition 👍

**Comment by MichealReed** [comment]
This looks like it's working great but my install step fails with 

```bash
CMake Error at src/dawn/native/cmake_install.cmake:51 (file):
  file INSTALL cannot find
  'gpu.cpp/third_party/local/dawn/build/Release/webgpu_dawn.dll":
  No error.
Call Stack (most recent call first):
  src/dawn/cmake_install.cmake:42 (include)
  cmake_install.cmake:80 (include)
```

I'm not sure I've seen a webgpu_dawn.dll so not sure why it's looking for that as a part of this install.

**Comment by kagurazaka-ayano** [comment]
> This looks like it's working great but my install step fails with
> 
> ```shell
> CMake Error at src/dawn/native/cmake_install.cmake:51 (file):
>   file INSTALL cannot find
>   'gpu.cpp/third_party/local/dawn/build/Release/webgpu_dawn.dll":
>   No error.
> Call Stack (most recent call first):
>   src/dawn/cmake_install.cmake:42 (include)
>   cmake_install.cmake:80 (include)
> ```
> 
> I'm not sure I've seen a webgpu_dawn.dll so not sure why it's looking for that as a part of this install.

seems like the problem with windows build. Unfortunately I don't have a windows device so I might not be able to test this. 

Or is it because find_library works in a different way on windows?

**Comment by kagurazaka-ayano** [comment]
closing since pr has been merged

**Comment by kagurazaka-ayano** [comment]
Reopening since there is a breaking change on [WebGPU-distribution](https://github.com/eliemichel/WebGPU-distribution) side (they removed branch dawn).

In cmake/webgpu.cmake:

```
option(WEBGPU_TAG "WebGPU distribution tag to use")
if(NOT WEBGPU_TAG)
    set(WEBGPU_TAG "dawn")
endif()
message(STATUS "Using WebGPU distribution tag: ${WEBGPU_TAG}")

if(WEBGPU_TAG STREQUAL "dawn")
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DWEBGPU_BACKEND_DAWN")
    # use specific commit set(WEBGPU_TAG
    # "1025b977e1927b6d0327e67352f90feb4bcf8274") set(WEBGPU_TAG
    # "acf972b7b909f52e183bdae3971b93bb13d4a29e")
    # add_compile_options(-UABSL_INTERNAL_AT_LEAST_CXX20) set(CMAKE_CXX_FLAGS
    # "${CMAKE_CXX_FLAGS} -UABSL_INTERNAL_AT_LEAST_CXX20")
    message(STATUS "Using Dawn backend")
endif()

FetchContent_Declare(
    webgpu
    GIT_REPOSITORY ${WEBGPU_DIST_GIT_REPO}
    GIT_TAG ${WEBGPU_TAG}
    GIT_SHALLOW TRUE)
FetchContent_MakeAvailable(webgpu)
```

I think this is the place where the webgpu is cloned.
